c10::KernelFunction

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26434 Move Generator ops to c10
* #26273 Enable creation of boxing wrappers for some aten operators
* #26272 Add two levels to use_c10_dispatcher
* **#26337 c10::KernelFunction**


- Factor out boxing and unboxing functionality from the c10 dispatcher into a c10::KernelFunction class
- Move that class and everything else it depends on into ATen/core/boxing
- This also allows us to get rid of c10::KernelCache. Instead, we now store a pointer to the unboxed functor in c10::KernelFunction.
- We're also getting rid of the DispatchTableEntry struct and instead store KernelFunction directly.
- To make this work, we need to change the dispatcher calling API from Dispatcher::lookup().callBoxed/callUnboxed and OperatorEntry::lookup().callBoxed/callUnboxed to Dispatcher::callBoxed/callUnboxed and OperatorEntry::callBoxed/callUnboxed.

Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

smessmer committed:
[wip] c10::KernelFunction



Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -37,9 +37,7 @@ using KernelCacheCreatorFunction = std::function<std::unique_ptr<c10::KernelCach
-  /*not-nullable*/ KernelCacheCreatorFunction cache_creator_func;
-  void* unboxed_kernel_func; // can be nullptr, not all kernels have this
+  KernelFunction kernel;
 };
```
Do you still need DispatchTableEntry, in this case?

smessmer replied:
not really, I can store KernelFunction directly.

ezyang commented on aten/src/ATen/core/dispatch/KernelFunction.h:
```diff
@@ -2,15 +2,208 @@
+  void callBoxed(Stack* stack) const {
+    if (C10_UNLIKELY(boxed_kernel_func_ == nullptr)) {
+      if (unboxed_kernel_func_ == nullptr) {
+        TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callBoxed() on an uninitizliaed KernelFunction.");
```
nit: uninitialized

ezyang commented on aten/src/ATen/core/dispatch/KernelFunction.h:
```diff
@@ -2,15 +2,208 @@
+  , signature_hash_(signature_hash)
+  {}
+
+  std::shared_ptr<OperatorKernel> functor_;
```
What's the `functor_` here?

smessmer replied:
That's the replacement for the old cache mechanism. It stores the actual kernel. If the kernel is registered as a functor, that functor is stored in here. If a kernel is registered as a runtime function pointer, we create a functor storing that pointer and store that functor in here.

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/KernelFunction.h:
```diff
@@ -2,15 +2,207 @@
+    return makeFromUnboxedFunctor<AllowLegacyTypes>(
+      std::make_shared<detail::WrapRuntimeKernelFunctor<guts::decay_t<Lambda>>>(std::forward<Lambda>(lambda))
+    );
+  }
```
Is it possible to reduce the surface area of this API? It was very irritating to have to fix multiple dispatch tests three times, because there were three copy pasted versions of morally the same test.

smessmer replied:
I don't think so unfortunately. The paths for function schema inference, for example, are slightly different for lambdas vs functions, and while writing that code, it did happen to me from time to time that only the test case for one of the APIs failed because I had forgotten something. I feel safer with all APIs being fully tested.

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] c10::KernelFunction"


Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::KernelFunction"



- Factor out boxing and unboxing functionality from the c10 dispatcher into a c10::KernelFunction class
- Move that class and everything else it depends on into ATen/core/boxing
- This also allows us to get rid of c10::KernelCache. Instead, we now store a pointer to the unboxed functor in c10::KernelFunction.
- We're also getting rid of the DispatchTableEntry struct and instead store KernelFunction directly.
- To make this work, we need to change the dispatcher calling API from Dispatcher::lookup().callBoxed/callUnboxed and OperatorEntry::lookup().callBoxed/callUnboxed to Dispatcher::callBoxed/callUnboxed and OperatorEntry::callBoxed/callUnboxed.

Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,466 @@
+namespace detail {
+template<class Type>
+constexpr uint64_t hashType() {
+  return typeid(Type).hash_code()
```
Doesn't this imply RTTI?

ezyang replied:
I guess, because this is constexpr, technically it's not? But I am not so sure about the implications.

smessmer replied:
I'm not 100% sure either. I'm trusting CI here. If this turns out to be impossible, it's easy to disable - it's just for error checking.

dzhulgakov replied:
did you verify (diasm in gdb or something) that it indeed becomes a constexpr value?

smessmer replied:
shit it actually does not. Good call. I could use `CAFFE_KNOWN_TYPE`, but that wouldn't be constexpr either. I'll remove the error checking code for now. I have an idea of how to do with constexpr it in C++14, maybe I can port it to C++11, but that'll be another diff.

smessmer replied:
Update: My idea works with C++11, but only GCC >= 5. So seems we need to wait for that to happen.

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,466 @@
+constexpr uint64_t hashType() {
+  return typeid(Type).hash_code()
+      + 1000 * std::is_lvalue_reference<Type>::value
+      + 5000 * std::is_rvalue_reference<Type>::value
```
value is true or false right? There's no reason to have such a large step here?

smessmer replied:
I use larger steps because I can't assume anything about the distribution of typeids. It's very possible that compilers choose adjacent type ids for types, i.e. if one type has id `250` then another type will have id `251` and so on. So I can't just use the numbers adjacent to the type id for representing const, that would be a collision. If I go larger steps away, it is less likely to collide. It's still not perfect, but less likely.

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,466 @@
+// Note that there it is not perfect error checking, two different signatures
+// might have the same hash, but it is probably good enough.
+template<class FuncSignature>
+constexpr uint64_t hashFunctionSignature() {
```
Can't you just build a completely precise struct describing the function signature in question? Rather than doing an approximate hash...

smessmer replied:
I could but that would be expensive to compare at runtime. Each time you call KernelFunction::callUnboxed, it's going to compare the args you call it with with the args of the functor that was used to create the KernelFunction. Doing that arg-by-arg is expensive, but computing a 64bit hash at compile time and at runtime just comparing two 64bit values is cheap.

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,466 @@
+  }
+
+  /**
+   * Call the function in an unboxed way.
```
At risk of sounding like a broken record, this API is sooooooo big. ATenDispatch had only a single way of registering a function and it made working with the code a lot easier. Doe we *really* need all these ways of registering functions? Maybe in an ideal world we would have all of them, but can we do without some, without too much trouble?

smessmer replied:
Technically, the API surface is smaller than it looks like. Most of these factory functions just forward to the functor one, they don't access any class internals. So technically, the API surface for creating KernelFunction is only two methods, one for creating from a boxed function and one for creating from an unboxed functor.

smessmer replied:
Note also that we need functors + functions + lambdas because they're part of the public custom op API. Some of them (the `unboxedOnly` variants) are needed only temporarily while we're doing the ATen->c10 transition and can be deleted later.

ezyang replied:
> Technically, the API surface is smaller than it looks like. Most of these factory functions just forward to the functor one, they don't access any class internals. So technically, the API surface for creating KernelFunction is only two methods, one for creating from a boxed function and one for creating from an unboxed functor.

The big step in complexity is having the internal implementation implement functors instead of just function pointers. We've given a lot of evidence that boxed functors are not actually necessary, *except* for the case when you are autogenerating a boxed wrapper for a unboxed function, but even then, rather than a general functor concept, a simple, NON-user-visible API for this use case would have been sufficient.

> Note also that we need functors + functions + lambdas because they're part of the public custom op API

There isn't that much usage yet. That is even more of an argument we should remove functionality now, before it becomes a big deal to remove it later. For reference, https://pytorch.org/tutorials/advanced/torch_script_custom_ops.html#registering-the-custom-operator-with-torchscript only advertises a *single* way of registering operators. Can you imagine having to explain three different APIs to users, and when to use them? It's totally unnecessary.

smessmer replied:
This PR just refactors things into a better state, it doesn't change any functionality or semantics and I won't change functionality in here, the PR is already complex enough to understand when you know it just refactors. If you want to change the public API, that's a different discussion we need to have, but that is unrelated to this PR.

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,466 @@
+  // We use this indirection because many KernelFunctions are created
+  // at static initialization time but are created with functors that
+  // store Tensor and we can't call the Tensor() constructor at static
+  // initialization time yet (SIOF). So these register with a
```
Err, really?

ezyang replied:
(I am specifically confused about the assertion that many functors store tensor)

smessmer replied:
I think you're right. This statement was true in the caffe2 world but is likely not true anymore in the PyTorch world. I changed it to "because we want to allow functors to store Tensors".

dzhulgakov replied:
Even in C2 world it's not true - we wouldn't create an instance of operator at registration time and lifetime of operator would be managed by the invocation context, not centralized dispatcher. I honestly don't think there's a usecase for kernelFactory_ today (and in the earlier design it was not fully figured out either). Can't we just nuke the functorFactory and just declare that it's always functor?

smessmer replied:
So we're saying it's never ok for a kernel functor to store Tensors? I feel uneasy about that since it seems an obvious thing to store. For example, one of the first designs for how to handle constants in the lite interpreter was to register a custom constant operator on the fly for each constant in a model when we load that model. That constant operator kernel would need to store the Tensor value it represents. The lite interpreter design changed and we don't need it anymore, but storing tensors still seems like an obviously useful feature.

If it really turns out to be not used, we can easily change this later. The existence of `kernelFactory_` is well hidden behind the custom op API, users don't see it.

ezyang replied:
I'm advocating for an even more hardline stance: NO closures, except for generated boxed wrappers (which need to know the pointer of the thing they're wrapping).

> Removing it now would require some changes to the experimental caffe2 ops I exported to c10

Can we just delete them? :)

smessmer replied:
Yes we could. Same point as above though, that would be a change in functionality and is unrelated to this PR. Let's set up a separate meeting to discuss this and not block this PR on it.

smessmer committed:
Update on "c10::KernelFunction"



- Factor out boxing and unboxing functionality from the c10 dispatcher into a c10::KernelFunction class
- Move that class and everything else it depends on into ATen/core/boxing
- This also allows us to get rid of c10::KernelCache. Instead, we now store a pointer to the unboxed functor in c10::KernelFunction.
- We're also getting rid of the DispatchTableEntry struct and instead store KernelFunction directly.
- To make this work, we need to change the dispatcher calling API from Dispatcher::lookup().callBoxed/callUnboxed and OperatorEntry::lookup().callBoxed/callUnboxed to Dispatcher::callBoxed/callUnboxed and OperatorEntry::callBoxed/callUnboxed.

Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,466 @@
+template<class Head, class... Tail>
+struct hashTypeList_<guts::typelist::typelist<Head, Tail...>> final {
+  static constexpr uint64_t call(uint64_t index) {
+    return 1000000 * index * hashType<Head>() + hashTypeList_<guts::typelist::typelist<Tail...>>::call(index + 1);
```
shameless programming competition plug - this is a really bad hash function :) (not that it's too critical here)

you're adding all items with `index*100000*hashType`, so they can easily collide

you probably wanted to make multiplicative hash - in which case you need to multiply the tail. But in that case 1000000 is a really bad constant as it'd overflow 2^64 and very quickly ignore other args. You want a co-prime number with 2^64. Even then, for best results you should take hashes module some large prime number - it lowers changes of saturation from e.g. indices being divisible by 2.

i.e. `(a0 + c*((a1 + c*(a2 ....))%mod))%mod` where `c` is usually larger than a_i and `mod` is a large prime (you can skip large prime part as you don't have int128 multiplication).

you can also look at folly::hash_combine that does bit rotation magic, but it's a bit too much for this case :)

smessmer replied:
Good points. Ignoring this for now because I removed the error checking, but will consider this if I manage to reintroduce it somehow.

dzhulgakov commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,466 @@
+// might have the same hash, but it is probably good enough.
+template<class FuncSignature>
+constexpr uint64_t hashFunctionSignature() {
+#if 0 // Disabling because the CUDA compiler complains.
```
you can probably guard it on `defined(__CUDACC__)`

smessmer replied:
Unfortunately not because it would be inconsistent hashes when linking cuda with non-cuda. Also ignoring this for now because I removed this code.

dzhulgakov commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,466 @@
+      return (*func)(getFunctor_(), std::forward<Args>(args)...);
+    }
+
+    TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callUnboxedOnly() for a kernel that doesn't have an unboxed version.");
```
is it the only difference between callUnboxedOnly and callBoxed? why do you need two functions? is it some kind of perf optimization to simplify forwarding?

dzhulgakov replied:
ah, you want to avoid resolving the template for boxing, makes sense. (explanation here is clearer: https://github.com/pytorch/pytorch/pull/26273#issuecomment-531974955 )

smessmer committed:
Update on "c10::KernelFunction"



- Factor out boxing and unboxing functionality from the c10 dispatcher into a c10::KernelFunction class
- Move that class and everything else it depends on into ATen/core/boxing
- This also allows us to get rid of c10::KernelCache. Instead, we now store a pointer to the unboxed functor in c10::KernelFunction.
- We're also getting rid of the DispatchTableEntry struct and instead store KernelFunction directly.
- To make this work, we need to change the dispatcher calling API from Dispatcher::lookup().callBoxed/callUnboxed and OperatorEntry::lookup().callBoxed/callUnboxed to Dispatcher::callBoxed/callUnboxed and OperatorEntry::callBoxed/callUnboxed.

Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::KernelFunction"



- Factor out boxing and unboxing functionality from the c10 dispatcher into a c10::KernelFunction class
- Move that class and everything else it depends on into ATen/core/boxing
- This also allows us to get rid of c10::KernelCache. Instead, we now store a pointer to the unboxed functor in c10::KernelFunction.
- We're also getting rid of the DispatchTableEntry struct and instead store KernelFunction directly.
- To make this work, we need to change the dispatcher calling API from Dispatcher::lookup().callBoxed/callUnboxed and OperatorEntry::lookup().callBoxed/callUnboxed to Dispatcher::callBoxed/callUnboxed and OperatorEntry::callBoxed/callUnboxed.

Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,397 @@
+  void callBoxed(Stack* stack) const {
+    if (C10_UNLIKELY(boxed_kernel_func_ == nullptr)) {
+      if (unboxed_kernel_func_ == nullptr) {
+        TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callBoxed() on an uninitizliaed KernelFunction.");
```
nit: uninitialized

smessmer committed:
Update on "c10::KernelFunction"



- Factor out boxing and unboxing functionality from the c10 dispatcher into a c10::KernelFunction class
- Move that class and everything else it depends on into ATen/core/boxing
- This also allows us to get rid of c10::KernelCache. Instead, we now store a pointer to the unboxed functor in c10::KernelFunction.
- We're also getting rid of the DispatchTableEntry struct and instead store KernelFunction directly.
- To make this work, we need to change the dispatcher calling API from Dispatcher::lookup().callBoxed/callUnboxed and OperatorEntry::lookup().callBoxed/callUnboxed to Dispatcher::callBoxed/callUnboxed and OperatorEntry::callBoxed/callUnboxed.

Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

smessmer committed:
Update on "c10::KernelFunction"



- Factor out boxing and unboxing functionality from the c10 dispatcher into a c10::KernelFunction class
- Move that class and everything else it depends on into ATen/core/boxing
- This also allows us to get rid of c10::KernelCache. Instead, we now store a pointer to the unboxed functor in c10::KernelFunction.
- We're also getting rid of the DispatchTableEntry struct and instead store KernelFunction directly.
- To make this work, we need to change the dispatcher calling API from Dispatcher::lookup().callBoxed/callUnboxed and OperatorEntry::lookup().callBoxed/callUnboxed to Dispatcher::callBoxed/callUnboxed and OperatorEntry::callBoxed/callUnboxed.

Differential Revision: [D17416607](https://our.internmc.facebook.com/intern/diff/D17416607/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,397 @@
+        TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callBoxed() on an uninitialized KernelFunction.");
+      } else {
+        // TODO We want to introduce the invariant that all kernels must be callable in a boxed way, then this case should be impossible.
+        TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callBoxed() on a KernelFunction that can only be called with KernelFunction::callUnboxed().");
```
So, in the meantime, this should be a `TORCH_CHECK`, right?

smessmer replied:
I don't think so. This doesn't fail due to user error but if we mucked up codegen. KernelFunction is not called directly by users and I don't think they can trigger this by using public APIs. Can you give an explanation why it should be TORCH_CHECK?

ezyang replied:
I thought the TODO implied this was triggerable from userspace.

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,397 @@
+   */
+  void callBoxed(Stack* stack) const {
+    if (C10_UNLIKELY(boxed_kernel_func_ == nullptr)) {
+      if (unboxed_kernel_func_ == nullptr) {
```
As an organizational point, it would be nice for this error handling code to live in the cpp file rather than the header. There's no reason for it to live inline, and it would make certain types of refactoring/debugging work easier (as it would avoid a recompilation of the world)

smessmer replied:
Calling operators is a hot path that I want to be fully inlineable. I could move it to a KernelFunction_inl.h header if you think that's better? Let me know.

ezyang replied:
I'm talking about the error reporting code, not the calling code.

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,397 @@
+   *
+   * KernelFunction::callUnboxed() is generally better, since it will allow
+   * calling KernelFunctions that are backed by either boxed or unboxed
+   * kernels, but that one will not work for all types.
```
I don't understand, is this a compilation-time error? I would prefer if we have a single entry point `callUnboxed`, that just skips the fallback to boxed types if the types in question are not supported. This is how I implemented it in #26368

smessmer replied:
Yes, I chose a bit a different approach here. I ant the  compiler to show an error if you have an op with `use_c10_dispatcher: full` but the boxed fallback doesn't work. That means the `use_c10_dispatcher` flag is the only configuration setting you have to move to fully migrate ops over to the c10 dispatcher. If you have some SFINAE config to remove ops from creating boxed wrappers, that would add an additional configuration setting and make the system more complex.

ezyang replied:
I don't buy the argument that it's more complex. The reason I don't buy it is because #26368 is able to avoid a bunch of pieces which you have to do in this PR:

- There's no per-operator setting saying whether or not it's `full` or `unboxedOnly`. The templates figure it out automatically.
- There's no change to code generation, as is here. The templates figure it out automatically.
- The actual difference in the dispatcher is a single if statement https://github.com/pytorch/pytorch/pull/26368/files#diff-4f2e119446bc2edf856334581e5c70c9R217 testing if the boxed version is supported, and error if it is not

The only downside is that you don't get compile time errors in the way you've done here. But it's easy to turn all of the runtime errors into compile time errors, to see if you have 100% coverage: change `supports_boxed_fallback` to return unconditionally true.

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,397 @@
+
+    // TODO Remove this function once all kernels support a boxed variant
+
+    if (unboxed_kernel_func_ != nullptr) {
```
C10_LIKELY?

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,397 @@
+    // TODO Remove this function once all kernels support a boxed variant
+
+    if (unboxed_kernel_func_ != nullptr) {
+      using ActualSignature = Return (OperatorKernel*, Args...);
```
This signature looks weird to me. Why is the unboxed kernel taking an `OperatorKernel*` as an argument? (I.e., what does this enable you to implement?) For reference, unboxed FuncType in #26368 is just `Return(Args...)` which provides evidence that you don't need the `OperatorKernel*`

smessmer replied:
The autogenerated boxed wrappers need to call into the unboxed functor so they need a pointer to it. This replaces what `c10::KernelCache` did before, which was also passed in to the boxed kernel.

ezyang replied:
You're right!

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,397 @@
+    if (unboxed_kernel_func_ != nullptr) {
+      using ActualSignature = Return (OperatorKernel*, Args...);
+      ActualSignature* func = reinterpret_cast<ActualSignature*>(unboxed_kernel_func_);
+      return (*func)(getFunctor_(), std::forward<Args>(args)...);
```
Are you using the term functor here in an untraditional way? In C++, I expect functor to be a "function like object". But you don't call the functor, you call a function pointer, passing the functor as an argument. That's weird?

smessmer replied:
this is calling into the boxed kernel, passing in the functor so the boxing wrapper can call it if the boxed kernel is just a wrapper that calls into the unboxed functor.

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,397 @@
+  , unboxed_kernel_func_(unboxed_kernel_func)
+  {}
+
+  OperatorKernel* getFunctor_() const {
```
Why does this method have a trailing underscore? Why is it called "get functor" but returns an "operator kernel*"?

smessmer replied:
because `c10::OperatorKernel` is the public API all kernel functors have to inherit from in the registration API.

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,397 @@
+    (*boxed_kernel_func)(functor, &stack);
+
+    TORCH_INTERNAL_ASSERT(stack.size() == 1, "A boxed kernel should only push one return to the stack");
+    return std::move(stack[0]).to<Return>();
```
Isn't `stack[0]` already a temporary, so the `std::move` is not necessary?

smessmer replied:
no, `std::vector<T>::operator[]` returns a `T&`, not a `T`.

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -0,0 +1,397 @@
+  }
+};
+template<class... Args>
+struct boxAndCallBoxedFunc<void, Args...> final {
```
Man, I hate that `void` needs special casing here XD

smessmer replied:
yes, me too. `if constexpr` would fix this but that's C++17 only.

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -81,32 +27,42 @@ class OperatorEntry final {
-      return OpKernel(kernel.kernel_func, kernel.cache_creator_func, kernel.unboxed_kernel_func);
+  template<class Return, class... Args>
+  Return callUnboxed(TensorTypeId dispatchKey, Args... args) const {
+    // TODO Remove dispatchKey argument and instead infer dispatchKey from args...
```
As I think you are aware, `multi_dispatch_tensor_type_set` should work well for this.

smessmer replied:
yes, I designed it towards being used with that.

ezyang reviewed (approved):
This seems like an improvement. But I still think this code can be simplified more.

smessmer commented:
Going to land this. As summary, I think there's three main points we need to address in future:

- It doesn't do error checking for `callUnboxed()` yet. I'm currently working on moving our code base to GCC 5, which would allow me to add error checking using compile time type ids.
- The `kernelFactory_` is only needed if we want to allow kernel functors to store Tensors (or other things that can't be created at static initialization time).
- The API is quite wide because our custom op API offers a lot of ways of creating kernels. This could be narrowed if we're willing to break the custom op API.
